packages/calling-component-bindings/src/handlers/createHandlers.ts

```diff
@@ -1,426 +1,58 @@
-    options?: StartCallOptions
-  ) => Call | undefined;
-  onSelectMicrophone: (device: AudioDeviceInfo) => Promise<void>;
-  onSelectSpeaker: (device: AudioDeviceInfo) => Promise<void>;
-  onSelectCamera: (device: VideoDeviceInfo, options?: VideoStreamOptions) => Promise<void>;
-  onToggleMicrophone: () => Promise<void>;
-  onStartScreenShare: () => Promise<void>;
-  onStopScreenShare: () => Promise<void>;
-  onToggleScreenShare: () => Promise<void>;
-  onHangUp: (forEveryone?: boolean) => Promise<void>;
-  /* @conditional-compile-remove(PSTN-calls) */
-  onToggleHold: () => Promise<void>;
-  /* @conditional-compile-remove(PSTN-calls) */
-  onAddParticipant(participant: CommunicationUserIdentifier): Promise<void>;
-  /* @conditional-compile-remove(PSTN-calls) */
-  onAddParticipant(participant: PhoneNumberIdentifier, options: AddPhoneNumberOptions): Promise<void>;
-  onCreateLocalStreamView: (options?: VideoStreamOptions) => Promise<void | CreateVideoStreamViewResult>;
-  onCreateRemoteStreamView: (
-    userId: string,
-    options?: VideoStreamOptions
-  ) => Promise<void | CreateVideoStreamViewResult>;
-  onRemoveParticipant(userId: string): Promise<void>;
-  /* @conditional-compile-remove(PSTN-calls) */
-  onRemoveParticipant(participant: CommunicationIdentifier): Promise<void>;
-  onDisposeRemoteStreamView: (userId: string) => Promise<void>;
-  onDisposeLocalStreamView: () => Promise<void>;
-  /* @conditional-compile-remove(dialpad) */ /* @conditional-compile-remove(PSTN-calls) */
-  onSendDtmfTone: (dtmfTone: DtmfTone) => Promise<void>;
-  /* @conditional-compile-remove(call-readiness) */
-  askDevicePermission: (constrain: PermissionConstraints) => Promise<void>;
-};
+export type CallTypeOf<
+  AgentType extends CallAgent | /* @conditional-compile-remove(teams-identity-support) */ TeamsCallAgent
+> = AgentType extends CallAgent ? Call : never | /* @conditional-compile-remove(teams-identity-support) */ TeamsCall; // remove "never" type when move to stable
 
 /**
- * @private
+ * @public
  */
-export const areStreamsEqual = (prevStream: LocalVideoStream, newStream: LocalVideoStream): boolean => {
-  return !!prevStream && !!newStream && prevStream.source.id === newStream.source.id;
-};
+export type CallHandlersOf<
+  AgentType extends CallAgent | /* @conditional-compile-remove(teams-identity-support) */ TeamsCallAgent
+> = AgentType extends CallAgent
+  ? CallingHandlers
+  : never | /* @conditional-compile-remove(teams-identity-support) */ TeamsCallingHandlers; // remove "never" type when move to stable
 
 /**
- * Create the default implementation of {@link CallingHandlers}.
- *
- * Useful when implementing a custom component that utilizes the providers
- * exported from this library.
- *
  * @public
  */
 export const createDefaultCallingHandlers = memoizeOne(
-  (
+  <AgentType extends CallAgent | /* @conditional-compile-remove(teams-identity-support) */ TeamsCallAgent = CallAgent>(
```

Comment:
Are the types `CallTypeOf` and `CallHandlersOf` _only_ introduced to support a generic `createDefaultCallingHandlers`?

In that case, I'd say we consider an alternative where these (difficult to understand) types are not introduced at the cost of having to know which of two functions `createDefaultCallingHandlers` vs `createDefaultTeamsCallingHandlers` must be called. This function should anyway not be needed frequently by devs.

iiuc, `useHandlers` will continue to work seamlessly for both types of calls and does not need new generic types to be introduced.

Reply:
You are 100% right! I've refactored the logic to my solution 2 - which doesn't require expose such a generic type